src/state.ts

```diff
@@ -134,100 +135,66 @@ function getProjectWsFolder(): vscode.WorkspaceFolder | undefined {
-    const docPath = doc && path.dirname(doc.uri.fsPath);
-
-    let currentPath =
-      util.isDefined(docPath) && docPath !== '.' ? docPath : workspaceFolder.uri.fsPath;
-
-    let previousPath = '';
-
-    do {
-      for (const projectFile of projectFileNames) {
-        const fullPath = path.resolve(currentPath, projectFile);
-        if (fs.existsSync(fullPath)) {
-          rootPath = currentPath;
-          break;
-        }
-      }
-      if (currentPath === rootPath) {
-        break;
-      }
-      previousPath = currentPath;
-      currentPath = path.resolve(currentPath, '..');
-    } while (currentPath !== previousPath);
-
-    // at least be sure the the root folder contains a
-    // supported project.
-    for (const projectFile of projectFileNames) {
-      const fullPath = path.resolve(rootPath, projectFile);
-      if (fs.existsSync(fullPath)) {
-        setStateValue(PROJECT_DIR_KEY, rootPath);
-        setStateValue(PROJECT_DIR_URI_KEY, vscode.Uri.file(rootPath));
-        return;
-      }
-    }
-  }
-  return;
-}
-
-async function findProjectRootUri(projectFileNames, doc, workspaceFolder): Promise<void> {
-  let searchUri = doc?.uri || workspaceFolder?.uri;
-  if (searchUri && !(searchUri.scheme === 'untitled')) {
-    let prev = null;
-    while (searchUri != prev) {
-      try {
-        for (const projectFile in projectFileNames) {
-          const u = vscode.Uri.joinPath(searchUri, projectFileNames[projectFile]);
-          try {
-            await vscode.workspace.fs.stat(u);
-            setStateValue(PROJECT_DIR_URI_KEY, searchUri);
-            return;
-          } catch {
-            // continue regardless of error
-          }
-        }
-      } catch (e) {
-        console.error(`Problems in search for project root directory: ${e}`);
-      }
-      prev = searchUri;
-      searchUri = vscode.Uri.joinPath(searchUri, '..');
-    }
+    console.log(workspaceFolder);
+    const candidatePaths = await findProjectRootPaths();
+    const closestRootPath = findClosestProjectRootPath(candidatePaths);
+    const projectRootPath = await pickProjectRootPath(candidatePaths, closestRootPath);
+    setStateValue(PROJECT_DIR_KEY, projectRootPath);
+    setStateValue(PROJECT_DIR_URI_KEY, vscode.Uri.file(projectRootPath));
   }
 }
 
+async function findProjectRootPaths() {
+  const projectFileNames: string[] = ['project.clj', 'shadow-cljs.edn', 'deps.edn'];
+  const projectFilesGlob = `**/{${projectFileNames.join(',')}}`;
+  const excludeDirsGlob = `**/{${config.getConfig().projectRootsSearchExclude.join(',')}}`;
+  const t0 = new Date().getTime();
+  const candidateUris = await vscode.workspace.findFiles(projectFilesGlob, excludeDirsGlob, 10000);
+  console.debug('glob took', new Date().getTime() - t0, 'ms');
+  const projectFilePaths = candidateUris.map((uri) => path.dirname(uri.fsPath));
+  const candidatePaths = [...new Set(projectFilePaths)];
+  console.log(candidatePaths);
```

Comment:
It might be cool to make this `console.log({candidatePaths})` so the log entry contents are labeled?

Reply:
TIL.

Wasn't intending to keep this logging in there, though. Except the timing of the glob, which could be good to be able to ask users about for a while.